lib/src/solver/solve_suggestions.dart

```diff
@@ -0,0 +1,214 @@
+import 'package:pub_semver/pub_semver.dart';
+
+import '../command_runner.dart';
+import '../entrypoint.dart';
+import '../flutter_releases.dart';
+import '../io.dart';
+import '../lock_file.dart';
+import '../package.dart';
+import '../package_name.dart';
+import '../pubspec_utils.dart';
+import '../solver.dart';
+import '../system_cache.dart';
+import 'incompatibility.dart';
+import 'incompatibility_cause.dart';
+
+/// Looks through the root-[incompability] of a solve-failure and tries to see if
+/// the conflict could resolved by any of the following suggestions:
+/// * An update of the current SDK.
+/// * Any single change to a package constraint.
+/// * Removing the bounds on all constraints, changing less than 5 dependencies.
+/// * Running `pub upgrade --major versions`.
+///
+/// Returns a formatted list of suggestions, or the empty String if no
+/// suggestions were found.
+Future<String?> suggestResolutionAlternatives(
+  Entrypoint entrypoint,
+  SolveType type,
+  Incompatibility incompatibility,
+  Iterable<String> unlock,
+  SystemCache cache,
+) async {
+  Future<_ResolutionSuggestion?> suggestSdkUpdate(SdkCause cause) async {
```

Comment:
Should we not make these top-level functions, and document what they do?

Reply:
We could create a:
```dart

class _FailedResolutionContext {
  final Entrypoint entrypoint;
  final SolveType type;
  final Incompatibility incompatibility;
  final Iterable<String> unlock;
  final SystemCache cache;
}

```

So we don't have to repeat all the arguments...

Reply:
Did that, moved the local functions to be methods on the context.

I keep the incompatibility out - I have plans to also make suggestions for `dart pub upgrade` for the case where things resolve, but you don't end up on the latest.